pkg/storage/wal/segment.go

```diff
@@ -246,10 +250,48 @@ func (b *SegmentWriter) Reset() {
 		streamSegmentPool.Put(s)
 	}
 	b.streams = make(map[streamID]*streamSegment, 64)
-	b.buf1.Reset()
 	b.inputSize = 0
 }
 
+func (b *SegmentWriter) ToReader() (io.ReadSeekCloser, error) {
+	// snappy compression rate is ~5x , but we can not predict it, so we need to allocate bigger buffer to avoid allocations
+	buffer := encodedWalSegmentBufferPool.Get(int(b.inputSize / 3))
+	_, err := b.WriteTo(buffer)
+	if err != nil {
+		return nil, fmt.Errorf("failed to write segment to create a reader: %w", err)
+	}
+	return NewEncodedSegmentReader(buffer), nil
+}
+
+var (
+	_ io.ReadSeekCloser = &EncodedSegmentReader{}
+)
+
+type EncodedSegmentReader struct {
+	delegate       io.ReadSeeker
+	encodedContent *bytes.Buffer
+}
+
+func NewEncodedSegmentReader(encodedContent *bytes.Buffer) *EncodedSegmentReader {
+	return &EncodedSegmentReader{
+		encodedContent: encodedContent,
+		delegate:       bytes.NewReader(encodedContent.Bytes()),
+	}
+}
+
+func (e EncodedSegmentReader) Read(p []byte) (n int, err error) {
+	return e.delegate.Read(p)
+}
+
+func (e EncodedSegmentReader) Seek(offset int64, whence int) (int64, error) {
+	return e.delegate.Seek(offset, whence)
+}
+
+func (e EncodedSegmentReader) Close() error {
+	encodedWalSegmentBufferPool.Put(e.encodedContent)
```

Comment:
you should protect double `Put` by setting it to nil here.

Reply:
done